File: Backtracking/DFS/46Permutatiton.py
# DFS, 每一层会删掉第一个element, 直到leaf之后一个数字, 然后回溯, 重新加入删掉的数字
# 在当前level时, 它包括了所有leaf的结果, 也就是所有可能的排列
# 所以再回到上一层时, 只需要加上当时删掉的数字即可

from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 0: return [[]]
        res = []

        for _ in range(len(nums)):
            cur = nums.pop(0)
            perms = self.permute(nums)

            for p in perms:
                p.append(cur)
            
            nums.append(cur)

            res.extend(perms)

        return res

File: Backtracking/DFS/46Permutatiton.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, res, 0, new ArrayList<>());
        return res;
    }

    private void dfs(int[] nums, List<List<Integer>> res, int index, List<Integer> cur) {
        if (index == nums.length) {
            res.add(new ArrayList<>(cur));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (cur.contains(nums[i])) {
                continue;
            }
            cur.add(nums[i]);
            dfs(nums, res, index + 1, cur);
            cur.remove(cur.size() - 1);
        }
    }
}